src/utils/async.js

```diff
@@ -1,5 +1,47 @@
 /* @flow strict-local */
-import { isClientError } from '../api/apiErrors';
+
+export class TimeoutError extends Error {
+  name = 'TimeoutError';
+}
+
+/**
+ * Time-out a Promise after `timeLimitMs` has passed.
+ *
+ * Returns a new Promise with the same outcome as `promise`, if
+ * `promise` completes in time.
+ *
+ * If `promise` does not complete before `timeLimitMs` has passed,
+ * `onTimeout` is called, and its outcome is used as the outcome of
+ * the returned Promise.
```

Comment:
I like this term "outcome". It's a good way of conveying this should apply to completion and failure uniformly.